core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -296,58 +316,69 @@ private ConnectivityState getAggregatedState() {
       static final int MAX_ENTRIES = 1000;
 
       final Key<String> key;
-      final Map<String, Ref<Subchannel>> stickinessMap =
-          new LinkedHashMap<String, Ref<Subchannel>>() {
-            @Override
-            protected boolean removeEldestEntry(Map.Entry<String,Ref<Subchannel>> eldest) {
-              return size() > MAX_ENTRIES;
-            }
-          };
+      final ConcurrentMap<String, Ref<Subchannel>> stickinessMap =
+          new ConcurrentHashMap<String, Ref<Subchannel>>();
 
-      final Map<Subchannel, Ref<Subchannel>> subchannelRefs =
-          new HashMap<Subchannel, Ref<Subchannel>>();
+      final Queue<Entry<String,Ref<Subchannel>>> lruQueue =
+              new ConcurrentLinkedQueue<Entry<String,Ref<Subchannel>>>();
 
       StickinessState(@Nonnull String stickinessKey) {
         this.key = Key.of(stickinessKey, Metadata.ASCII_STRING_MARSHALLER);
       }
 
       /**
-       * Returns the subchannel asscoicated to the stickiness value if available in both the
+       * Returns the subchannel associated to the stickiness value if available in both the
        * registry and the round robin list, otherwise associates the given subchannel with the
        * stickiness key in the registry and returns the given subchannel.
        */
       @Nonnull
-      synchronized Subchannel maybeRegister(
+      Subchannel maybeRegister(
           String stickinessValue, @Nonnull Subchannel subchannel, List<Subchannel> rrList) {
-        Subchannel existingSubchannel = getSubchannel(stickinessValue);
-        if (existingSubchannel != null && rrList.contains(existingSubchannel)) {
-          return existingSubchannel;
+        final Ref<Subchannel> newSubchannelRef = subchannel.getAttributes().get(STICKY_REF);
+        Ref<Subchannel> existingSubchannelRef = null;
+        while (true) {
+          if (existingSubchannelRef != null) {
+            if (stickinessMap.replace(stickinessValue, existingSubchannelRef, newSubchannelRef)) {
+              addToLruQueue(stickinessValue, newSubchannelRef, false);
```

Comment:
> LR-used is not required, it's totally okay for LR-created

SGTM but I do think that other things being equal, LR-used would be preferable.

Agree that if it's going to stay with LR-created, the queue may as well just be scoped to stickiness vals per your proposal rather than specific stickiness-val -> Subchannel mappings. I think it would introduce a small race condition where there could be an extra flip-flop between eviction and registration of a new mapping, but that's probably insignificant. I'll change it!

Could you confirm that using a guava cache isn't an option?

Reply:
A guava cache may increase the method count and make android proguard unhappy, need @ericgribkoff to confirm.

Reply:
Discuss with @ericgribkoff, Android does not pull RRLB dependency by default, so currently  android proguard CI will not fail if guava cache is used here. According to https://github.com/google/guava/wiki/Android , they recommend against using guava cache.

Generally guava cache is way powerful than needed here.

Reply:
I can see that it would be more powerful/heavy than needed in the Android case since that wouldn't have the same concurrency concerns (Android's `LruCache` uses course locking and does not support weak entries). I'd imagine client-side load balancing in general would be unusual in the Android context (presumably that's why it doesn't include RRLB).

But I still think it's a very good fit in the general case - i.e. a concurrent size or age bounded LRU map with weak values. So maybe it's still a possibility, WDYT?

Reply:
> client-side load balancing in general would be unusual in the Android context.

I agree

>But I still think it's a very good fit in the general case - i.e. a concurrent size or age bounded LRU map with weak values. So maybe it's still a possibility, WDYT?

Worth a try. Let's get the current PR in first.